server/datastore/mysql/locks.go

```diff
@@ -0,0 +1,76 @@
+package mysql
+
+import (
+	"time"
+)
+
+func (d *Datastore) Lock(name string, owner string, expiration time.Duration) (bool, error) {
+	lockObtainers := []func(string, string, time.Duration) (int64, error){
+		d.extendLockIfAlreadyAcquired,
+		d.overwriteLockIfExpired,
+		d.createLock,
+	}
+
+	for _, lock := range lockObtainers {
+		rowsAffected, err := lock(name, owner, expiration)
+		if err != nil {
+			return false, err
+		}
+		if rowsAffected > 0 {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (d *Datastore) createLock(name string, owner string, expiration time.Duration) (int64, error) {
+	res, err := d.db.Exec(
+		`INSERT IGNORE INTO locks (name, owner, expires_at) VALUES (?, ?, ?)`,
+		name, owner, time.Now().Add(expiration),
+	)
```

Comment:
Everything besides this portion looks the same in each function... Maybe these functions can be simplified and the rest of the checks can be moved into the loop above?